### PR TITLE
feat(generator): add constructor-based mapping support

### PR DESCRIPTION
**Pull request by j-d-ha**
# 🚀 Pull Request

## 📋 Summary

This PR adds comprehensive support for constructor-based mapping in DynamoMapper, enabling the source generator to deserialize DynamoDB items into C# records and classes with parameterized constructors. This is a major enhancement that expands the types of domain models that can be used with DynamoMapper.

### Key Features Added

1. **Constructor-Based Deserialization**: The generator now analyzes constructors and intelligently selects the best one for deserialization based on parameter-to-property matching.

2. **`[DynamoMapperConstructor]` Attribute**: Allows explicit constructor selection when multiple constructors are available.

3. **Primary Constructor Support**: Full support for C# 12+ primary constructors on records and classes.

4. **Init-Only Properties**: Support for properties with `init` accessors that can only be set during object construction.

5. **Hybrid Initialization**: Handles models that use both constructor parameters and settable properties, with smart property initialization mode detection.

6. **Comprehensive Diagnostics**: Added new compile-time diagnostics (DM0003-DM0007) to guide developers when constructor mapping issues are detected.

### Changes by Component

**Generator Enhancements**:
- `ConstructorSelector`: New component that analyzes and selects the best constructor for deserialization
- `MemberAnalyzer`: Shared logic for analyzing both properties and constructor parameters
- Enhanced `ModelClassInfo` with constructor metadata and initialization tracking
- Updated Scriban template to generate constructor-based `FromItem` methods

**Runtime API**:
- Added `DynamoMapperConstructorAttribute` to mark preferred constructors

**Testing**:
- Added 11 comprehensive snapshot tests covering various constructor scenarios
- Tests cover: primary constructors, attributed constructors, init-only properties, hybrid models, and error cases

**Documentation**:
- Updated `CLAUDE.md` and `AGENTS.md` with xUnit v3 test commands
- Added constructor-based examples in `DynamoMapper.MapperConstructor` project

### Technical Details

The constructor selection logic prioritizes:
1. Constructors explicitly marked with `[DynamoMapperConstructor]`
2. Primary constructors (for records/classes)
3. The constructor with the most parameters that can be satisfied by mapped properties

Property initialization modes are determined per-property:
- `ConstructorParameter`: Property is set via constructor parameter
- `DirectAssignment`: Property is set after construction (regular or `init` setter)
- `ReadOnly`: Property is read-only and only set via constructor

---

## ✅ Checklist

- [x] My changes build cleanly
- [x] I've added/updated relevant tests (11 new snapshot tests)
- [x] I've added/updated documentation (CLAUDE.md, AGENTS.md, examples)
- [x] I've followed the coding style for this project
- [x] I've tested the changes locally (all tests passing)

---

## 🧪 Related Issues or PRs

This PR implements a core Phase 1 feature for DynamoMapper, enabling it to work with a much broader range of domain models beyond simple POCOs with parameterless constructors.

---

## 💬 Notes for Reviewers

**Areas to Focus On**:

1. **Constructor Selection Logic** (`ConstructorSelector.cs`): The algorithm for choosing the best constructor and matching parameters to properties. Pay attention to the name matching logic (exact, case-insensitive, underscore prefix handling).

2. **Property Initialization Modes**: The logic in `ModelClassInfo.cs` that determines whether each property should be set via constructor parameter or direct assignment.

3. **Template Changes** (`Mapper.scriban`): The new `FromItem` generation logic that handles constructor invocation with parameter mapping.

4. **Diagnostics**: The new error messages (DM0003-DM0007) should be clear and actionable for developers.

5. **Test Coverage**: Review the snapshot tests in `ConstructorVerifyTests.cs` to ensure all edge cases are covered.

**Known Limitations**:
- Constructor parameters must match property names (case-insensitive, with underscore prefix tolerance)
- Circular dependencies between constructor parameters are not detected
- Performance of constructor selection for classes with many constructors has not been profiled

**Follow-up Work**:
- Performance profiling of constructor selection logic for classes with many properties
- Consider caching constructor analysis results more aggressively
- Potential enhancement: Support for factory methods as an alternative to constructors

### CODE REVIEW

**Review by ncipollina — CHANGES_REQUESTED** [review]
I stopped my review after my comment. Let's discuss.

File: global.json
Comment (ncipollina) [inline global.json]: ❓ Why did you add this?

File: global.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "sdk": {
+    "rollForward": "latestMinor",
+    "version": "10.0.101"
```
Comment (ncipollina) [inline global.json]: 🔄 this should be:
```suggestion
    "version": "10.0.102"
```

File: global.json
Comment (j-d-ha) [reply global.json]: Yes!
